File: python_src/zeabus/math/quaternion.py
#!/usr/bin/env python
# FILE			: quaternion.py
# AUTHOR		: K.Supasan , W.Natthaphat
# CREATE ON		: 2019, June 05 (UTC+0)
# MAINTAINER	: K.Supasan

# README
#   This code have inspiration to create from use python quaternion to distribute data between
#       parent frame and child frame
#   In the order of quaternion we will use standard follow ref1
#   Tuple of quaternion will show by order is x y z w

# REFERENCE
#   ref01 : http://docs.ros.org/melodic/api/tf/html/python/transformations.html#module-tf.transform
#   ref02 : https://docs.python.org/2/tutorial/classes.html
#   ref03 : https://www.geeksforgeeks.org/operator-overloading-in-python/
#   ref04 : http://www.chrobotics.com/library/understanding-quaternions

import math
import numpy
from . import general
from tf import transformations as tf_handle

class Quaternion :

    # I use vector because quaternion is subset of vector but quaternion is vector in
    #   4 Dimension

    # We will init quaternion at roll pitch yaw is 0 radian
    def __init__ (self , quaternion = ( 0 , 0 , 0 , 1 ) ):
        self.vector = quaternion
       
    # function set_quaternion use to set by input tuple of quaternion
    def set_quaternion (self , quaternion):
        self.vector = quaternion

    # We follow order rotation is yaw pitch roll.
    # This order will apply use in system of zeabus team
    def set_euler (self , yaw , pitch , roll):
        temp = tf_handle.quaternion_from_euler( yaw, pitch, roll, axes='rzyx')
        self.set_quaternion( temp )

    # function get_euler use to find euler from quaternion 
    def get_euler (self):
        temp = tf_handle.euler_from_quaternion( self.vector , axes='rzyx')
        result = ( general.bound_radian( temp[0] )
            , general.bound_radian( temp[1])
            , general.bound_radian( temp[2]) )
        return temp

    def print_quaternion( self , name ):
        print( "{:s} is {:6.3f} {:6.3f} {:6.3f} {:6.3f}".format( name
            , self.vector[0] 
            , self.vector[1] 
            , self.vector[2]
            , self.vector[3] ) )

    # function inverse of quaternion inverse of quaternion is conjugate all imginary part
    def inverse(self):
        q = Quaternion()
        q.set_quaternion( ( -self.vector[0], -self.vector[1], -self.vector[2], self.vector[3] ) )
        return q

    def normalize(self):
        temp = tf_handle.unit_vector( self.vector )
        self.set_quaternion( temp )

    # Function this use to rotation data
    def rotation(self, quaternion ):
        temp = Quaternion()
        temp.set_quaternion( quaternion )
        result = self * temp * self.inverse()
        return result

    # Function this use to inverse rotation data
    def inverse_rotation( self, quaternion ):
        temp = Quaternion()
        temp.set_quaternion( quaternion )
        result = self.inverse() * temp * self
        return result

    # This use in pattern of multiple between object quaternion
    def __mul__(self, other_quatanion):
        temp = tf_handle.quaternion_multiply( self.vector, other_quatanion.vector )
        q = Quaternion()
        q.set_quaternion(temp)
        return q

File: static_src/zeabus/math/quaternion.cpp
// FILE			: quaternion.cpp
// AUTHOR		: K.Supasan
// CREATE ON	: 2020, Febuary 25 (UTC+0)
// MAINTAINER	: K.Supasan

// MACRO DETAIL

// README

// REFERENCE

// MACRO SET

// MACRO CONDITION

#include    <zeabus/math/quaternion.hpp>

namespace zeabus
{

namespace math
{
    void get_euler( const tf::Quaternion source,
            double* roll , double* pitch , double* yaw )
    {
        *yaw = atan2( 2 * ( source.x() * source.y() + source.w() * source.z() ),
                pow( source.w() , 2 ) + pow( source.x() , 2 ) - 
                        pow( source.y() , 2 ) - pow( source.z() , 2 )  );
        *pitch = atan2( -2.0*( source.x() * source.z() - source.w() * source.y() ),
                sqrt( 1.0 - pow( 2.0*( source.x() * source.z() - source.w() * source.y() ) , 2 ) ) );
        *roll = atan2( 2.0 * ( source.y() * source.z() + source.w() * source.x() ) ,
                pow( source.w() , 2 ) - pow( source.x() , 2 ) - 
                        pow( source.y() , 2 ) + pow( source.z() , 2 )  );
    }    

    void get_euler( const geometry_msgs::Quaternion source,
            double* roll , double* pitch , double* yaw )
    {
        *yaw = atan2( 2 * ( source.x * source.y + source.w * source.z ),
                pow( source.w , 2 ) + pow( source.x , 2 ) - 
                        pow( source.y , 2 ) - pow( source.z , 2 )  );
        *pitch = atan2( -2.0*( source.x * source.z - source.w * source.y ),
                sqrt( 1.0 - pow( 2.0*( source.x * source.z - source.w * source.y ) , 2 ) ) );
        *roll = atan2( 2.0 * ( source.y * source.z + source.w * source.x ) ,
                pow( source.w , 2 ) - pow( source.x , 2 ) - 
                        pow( source.y , 2 ) + pow( source.z , 2 )  );
    }

    void rotation( const tf::Quaternion& quaternion , boost::array< double , 3 >* vector)
    {
        tf::Quaternion temp_quaternion( vector->at( 0 ), vector->at( 1 ) , vector->at( 2 ) , 0 );
        temp_quaternion = quaternion.inverse() * temp_quaternion * quaternion;
        vector->at( 0 ) = temp_quaternion.x();
        vector->at( 1 ) = temp_quaternion.y();
        vector->at( 2 ) = temp_quaternion.z();
    } 

    void inv_rotation( const tf::Quaternion& quaternion , boost::array< double , 3 >* vector )
    {
        tf::Quaternion temp_quaternion( vector->at( 0 ), vector->at( 1 ) , vector->at( 2 ) , 0 );
        temp_quaternion = quaternion * temp_quaternion * quaternion.inverse();
        vector->at( 0 ) = temp_quaternion.x();
        vector->at( 1 ) = temp_quaternion.y();
        vector->at( 2 ) = temp_quaternion.z();
    }

    void rotation( const tf::Quaternion& quaternion , double* vector)
    {
        tf::Quaternion temp_quaternion( vector[0], vector[1] , vector[2] , 0 );
        temp_quaternion = quaternion.inverse() * temp_quaternion * quaternion;
        vector[0] = temp_quaternion.x();
        vector[1] = temp_quaternion.y();
        vector[2] = temp_quaternion.z();
    } 

    void inv_rotation( const tf::Quaternion& quaternion , double* vector )
    {
        tf::Quaternion temp_quaternion( vector[0], vector[1] , vector[2] , 0 );
        temp_quaternion = quaternion * temp_quaternion * quaternion.inverse();
        vector[0] = temp_quaternion.x();
        vector[1] = temp_quaternion.y();
        vector[2] = temp_quaternion.z();
    }

    void rotation( const tf::Quaternion& quaternion , tf::Vector3* vector)
    {
        tf::Quaternion temp_quaternion( vector->x(), vector->y() , vector->z() , 0 );
        temp_quaternion = quaternion.inverse() * temp_quaternion * quaternion;
        *vector = tf::Vector3( temp_quaternion.x() , temp_quaternion.y() , temp_quaternion.z() );
    } 

    void inv_rotation( const tf::Quaternion& quaternion , tf::Vector3* vector )
    {
        tf::Quaternion temp_quaternion( vector->x(), vector->y() , vector->z() , 0 );
        temp_quaternion = quaternion * temp_quaternion * quaternion.inverse();
        *vector = tf::Vector3( temp_quaternion.x() , temp_quaternion.y() , temp_quaternion.z() );
    }

    void get_quaternion( const double roll , const double pitch , const double yaw,
            tf::Quaternion* quaternion )
    {
        double sp = sin( pitch / 2 ); // sin( pitch / 2 )
        double cp = cos( pitch / 2 ); // cos( pitch / 2 )
        double sr = sin( roll / 2 ); // sin( roll / 2 )
        double cr = cos( roll / 2 ); // cos( roll / 2 )
        double sy = sin( yaw / 2 ); // sin( yaw / 2 )
        double cy = cos( yaw / 2 ); // cos( yaw / 2 )
        *quaternion = tf::Quaternion(
            -1.0 * sy * sp * cr + cy * cp * sr, // x
            +1.0 * sy * cp * sr + cy * sp * cr, // y
            +1.0 * sy * cp * cr - cy * sp * sr, // z
            +1.0 * sy * sp * sr + cy * cp * cr  // w
        );
    }
} // namespace math

} // namespace zeabus
